From c6ef0eeed512f0b3cfe343a3407238cf414a480e Mon Sep 17 00:00:00 2001
From: Venkatasaiprudhvi Sannidhi <vsannidhi@maxlinear.com>
Date: Tue, 27 May 2025 17:52:12 +0530
Subject: [PATCH] WLANRTSYS-89055 Fix for DUT Hung observed during reboot
 stability test.

Issue: On issuing "reboot -f" as part of platform stability test for 100
iterations, the AP goes into unresponsive state.

Root cause: In mac80211, during the interface unregistration procedure while bringing
down the VAPs, the lock for accessing wiphy is being held and then the
net devices unregistration from the unreg_list takes place. The DUT gets
stuck in the mac80211 api "unregister_netdevice_many".

Fix: Acquire the wiphy lock only when accessing the wdev_list.

Unit test: With dry run image, checked that 100 iterations of reboot -f
passed.
---
 net/mac80211/iface.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/net/mac80211/iface.c b/net/mac80211/iface.c
index 64a691d0b195..d73fd68af245 100644
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -2349,8 +2349,6 @@ void ieee80211_remove_interfaces(struct ieee80211_local *local)
 	 */
 	cfg80211_shutdown_all_interfaces(local->hw.wiphy);
 
-	wiphy_lock(local->hw.wiphy);
-
 	WARN(local->open_count, "%s: open count remains %d\n",
 	     wiphy_name(local->hw.wiphy), local->open_count);
 
@@ -2371,6 +2369,7 @@ void ieee80211_remove_interfaces(struct ieee80211_local *local)
 
 	unregister_netdevice_many(&unreg_list);
 
+	wiphy_lock(local->hw.wiphy);
 	list_for_each_entry_safe(sdata, tmp, &wdev_list, list) {
 		bool netdev = sdata->dev;
 
-- 
2.17.1

